watertap/core/zero_order_sido.py

```diff
@@ -359,32 +312,114 @@ def calculate_scaling_factors(self):
+            self, parameter, data, index=None, use_default_removal=False):
+        """
+        General method for setting parameter values from a dict of data
+        returned from a database.
 
-        for t, v in self.byproduct_temperature_equality.items():
-            iscale.constraint_scaling_transform(
-                v, iscale.get_scaling_factor(
-                    self.properties_in[t].temperature,
-                    default=1e-2,
-                    warning=False))  # would just be a duplicate of above
+        Args:
+            parameter - a Pyomo Var to be fixed to value from database
+            data - dict of parameter values from database
+            index - (optional) index to fix if parameter is an IndexedVar
+            use_default_removal - (optional) indicate whether to use defined
+                                  default removal fraction if no specific value
+                                  defined in database
+
+        Returns:
+            None
+
+        Raises:
+            KeyError if values cannot be found for parameter in data dict
+
+        """
+
+        pname = parameter.parent_component().local_name
+
+        try:
+            pdata = data[pname]
+        except KeyError:
+            raise KeyError(
+                f"{self.name} - database provided does not contain an entry "
+                f"for {pname} for technology.")
+
+        if index is not None:
+            try:
+                pdata = pdata[index]
+            except KeyError:
+                if pname == "removal_mass_solute" and use_default_removal:
+                    try:
+                        pdata = data["default_removal_mass_solute"]
+                        index = "default"
+                    except KeyError:
+                        raise KeyError(
+                            f"{self.name} - database provided does not "
+                            f"contain an entry for {pname} with index {index} "
+                            f"for technology and no default removal was "
+                            f"specified.")
+                else:
+                    raise KeyError(
+                        f"{self.name} - database provided does not contain "
+                        f"an entry for {pname} with index {index} for "
+                        f"technology.")
+
+        try:
+            val = pdata["value"]
+        except KeyError:
+            raise KeyError(
+                f"{self.name} - no value provided for {pname} (index: "
+                f"{index}) in database.")
+        try:
+            units = getattr(pyunits, pdata["units"])
+        except KeyError:
+            raise KeyError(
+                f"{self.name} - no units provided for {pname} (index: "
+                f"{index}) in database.")
+
+        parameter.fix(val*units)
+        _log.info_high(f"{parameter.name} fixed to value {val} {str(units)}")
+
+    def set_recovery_and_removal(self, data, use_default_removal=False):
```

Comment:
Should this method be "private", since it's only to be called from derived classes?

Reply:
I originally made it private, but made it public so as not to discourage developers of derived classes from using it (less-experienced users might avoid using any private methods).

Reply:
Good point -- I never let an underscore stop me, but that's not everyone